src/utils/collection.rs

```diff
@@ -0,0 +1,236 @@
+    }
+
+    /// Add a split BTF in the current BtfCollection representation, reading a file.
+    pub fn add_split_btf_from_file<P: AsRef<Path>>(&mut self, path: P) -> Result<&mut Self> {
+        let name = match path.as_ref().file_name() {
+            Some(name) => match name.to_str() {
+                Some(s) => s.to_string(),
+                None => bail!("Invalid file name {:?}", name),
+            },
+            None => bail!(
+                "Could not get file name from path {}",
+                path.as_ref().display()
+            ),
+        };
+
+        if self.split.iter().any(|m| m.name == name) {
+            bail!("Split BTF with name {name} already present");
+        }
+
+        self.split.push(NamedBtf {
+            name,
+            btf: Btf::from_split_file(path, &self.base.btf)?,
+        });
+        Ok(self)
+    }
+
+    /// Add a split BTF in the current BtfCollection representation, reading a byte slice.
+    pub fn add_split_btf_from_bytes(&mut self, name: &str, bytes: &[u8]) -> Result<&mut Self> {
+        let name = name.to_string();
+        if self.split.iter().any(|m| m.name == name) {
+            bail!("Split BTF with name {name} already present");
+        }
+
+        self.split.push(NamedBtf {
+            name,
+            btf: Btf::from_split_bytes(bytes, &self.base.btf)?,
+        });
+        Ok(self)
+    }
+
+    /// Parse BTF objects stored in a directory and construct a BtfCollection
+    /// object, given a path to the directory and the filename of the base BTF file.
+    /// This is helpful for parsing /sys/kernel/btf for example.
+    pub fn from_dir<P: AsRef<Path>>(dir: P, base: &str) -> Result<BtfCollection> {
+        // First parse the base BTF information.
+        let mut sys_btf = BtfCollection::from_file(format!("{}/{base}", dir.as_ref().display()))?;
+
+        // Then loop over all split BTF files and parse them.
+        for file in fs::read_dir(dir.as_ref())? {
+            match file {
+                Ok(file) => {
+                    if file.file_name() == base {
+                        continue;
+                    }
+                    if let Ok(ft) = file.file_type() {
+                        if !ft.is_dir() {
+                            sys_btf.add_split_btf_from_file(file.path())?;
+                        }
+                    }
+                }
+                Err(e) => bail!("Error reading file from {}: {e}", dir.as_ref().display()),
+            }
+        }
+
+        Ok(sys_btf)
+    }
+
+    /// Find a list of BTF ids using their name as a key. Matching ids can be
+    /// found in multiple underlying BTF, thus this function returns a list of
+    /// tuples containing each a reference to `NamedBtf` (representing the BTF
+    /// where a match was found) and the id. Further lookups must be done using
+    /// the `Btf` object contained in the linked `NamedBtf` one.
+    pub fn resolve_ids_by_name(&self, name: &str) -> Result<Vec<(&NamedBtf, u32)>> {
```

Comment:
Servers can have many modules. However, it's likely that most of the lookups target the base. This function looks N+1 times on the same base file, which is quite a lot of completely redundant lookups.

I think we should at least have a function that returns on the first hit to avoid increasing the lookup time needlessly.

On the other hand, what use case needs all matches on a vector? What would a client do if there is more than one? How would the user know it's the type they are looking for? The only possible thing (I can think of) to do next would be to iterate through them and find the one coming from expected module.

If that's the case, wouldn't it be easier to let the user specify the named file (i.e: module name) where the lookup is expected to return a result?

Reply:
Sorry for the very long reply time /o\.


> > Servers can have many modules. However, it's likely that most of the lookups target the base. This function looks N+1 times on the same base file, which is quite a lot of completely redundant lookups.
> 
> I didn't wanted to modify our core objects, but you're right. One solution would be to have a crate-only helper in `Btf` to skip the base lookup? It does not make sense from an user perspective at the `Btf` level but it totally makes sense for any stacked construction like this, which then the crate provides to users.
> 

Yep, that would work.

> > I think we should at least have a function that returns on the first hit to avoid increasing the lookup time needlessly.
> 
> Or return an iterator and not a `Vec`? If you only use `next()` once you only lookup once.
> 

I like that idea.


> > On the other hand, what use case needs all matches on a vector? What would a client do if there is more than one? How would the user know it's the type they are looking for? The only possible thing (I can think of) to do next would be to iterate through them and find the one coming from expected module.
> > If that's the case, wouldn't it be easier to let the user specify the named file (i.e: module name) where the lookup is expected to return a result?
> 
> IMHO those are two different use cases. If you know where to find a given information, I agree an helper is missing to explicitly request the lookup to happen in a specific split BTF representation. If you don't know, then the current helpers can be used. In such case, yes, we could find more than one match and `btf-rs` cannot make the decision on its own on what to do. The user might ignore this, report a warning, or inspect the result.
> 
> One comment would be if an user know exactly where to look for an information it can use a `Btf` representation directly, whereas the full system representation can be used when it's not directly clear. But it also makes sense to do both, so to answer your comment, what about an additional helper to return a reference on a specific `NamedBtf`?
> 
> ```rust
> fn get_named_btf(&self, name: &str) -> Option<&NamedBtf> {}
> ```
> 
> Which can then be used directly as usual `Btf` objects. This way the construction is clear: `BtfCollection` operates on the full representation while `Btf` on a single one (split or not).
> 
> WDYT?

That would make it easier for users that know where to look for, yes.

Reply:
I added split-only helpers to avoid looking into the base BTF definition multiple times, and a `get_named_btf` helper to allow users to retrieve a simple split BTF definition.

I haven't implemented using an iterator as IMO this should be a broader work in btf-rs: if we add it (I think we should), we should add an iterator variant to every `resolve_<>s_by_<>` helper and not only in the collection.